spec private class (#9533)
scala/scala: spec/05-classes-and-objects.md

the validity and meaning of a modifier are as follows.

### `private`

> The `private` modifier can be used with any template definition or any definition or

Reviewer: "any template" is not accurate, local definitions can't be `private`. Maybe also say "top-level" here?

```
scala> def f = { private class C; 0 }
                 ^
       error: illegal start of statement (no modifiers allowed here)
```

Author: Maybe split it up then?

```
The `private` modifier can be used with any definition or declaration in a
template. Private members of a template can be accessed only from within the
directly enclosing template and its companion module or
[companion class](#object-definitions).

The `private` modifier is also valid for
[top-level](09-top-level-definitions.html#packagings) templates. Private
top-level templates can be accessed only from within the same package.
```

Reviewer: Yes I think that makes sense. I'd mark it as "Note" and only keep ``The `private` modifier is also valid for
[top-level](09-top-level-definitions.html#packagings) templates`` instead of repeating it here.

declaration in a template. Private members of a template can be accessed only from
within the directly enclosing template and its companion module or

> [companion class](#object-definitions). Private templates that are not template
> members themselves can be accessed only from within the same package.

Reviewer: ```suggestion
[companion class](#object-definitions). Private [top-level](09-top-level-definitions.html#packagings)
templates can be accessed only from within the same package.
```

Author: Aha! It was already specced there:

> A packaging package p { ds } injects all definitions ds in as members into the package whose qualified name is p.
Members of a package are called top-level definitions. _If a definition in ds is labeled private, it is visible only for
other members in the package._

Still it may be good to have it here too.

Reviewer: Ah, right. Here we're already narrowed down to the spec for members. So it's probably better to add a `Note` below instead of changing the actual spec.

the validity and meaning of a modifier are as follows.

### `private`
The `private` modifier can be used with any definition or declaration in a
template. Private members of a template can be accessed only from within the
directly enclosing template and its companion module or
[companion class](#object-definitions).

The `private` modifier is also valid for

> [top-level](09-top-level-definitions.html#packagings) templates.

Reviewer 2: This doesn't say what the semantics are; I think you deleted it in an edit.

One might imagine a top-level private class C is private to object C instead of enclosing private[p].

Reviewer: I suggested not to repeat the semantics here, as they are defined in the linked section.

Reviewer 2: OK thanks for clarifying. I always forget to click the link. Please don't assume all your readers are so lazy.
